File: shared/states/go/advance.go
package states

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"log/slog"
)

// IllegalStateTransition is returned when (from, trigger, outcome) has
// no matching edge in the runtime transition table. It carries enough
// context for observability without leaking the entire table.
type IllegalStateTransition struct {
	VideoID string
	From    State
	Trigger Trigger
	Outcome Outcome
}

// Error implements error. Format is stable; tests grep for it.
func (e *IllegalStateTransition) Error() string {
	return fmt.Sprintf(
		"illegal state transition for video %s: from=%s trigger=%s outcome=%s",
		e.VideoID, e.From, e.Trigger, e.Outcome,
	)
}

// ErrIllegalTransition is the sentinel callers compare against with
// errors.Is. The concrete *IllegalStateTransition value still carries
// the from/trigger/outcome detail.
var ErrIllegalTransition = errors.New("illegal state transition")

// Is satisfies errors.Is so callers can write
// `errors.Is(err, states.ErrIllegalTransition)` without unwrapping.
func (e *IllegalStateTransition) Is(target error) bool {
	return target == ErrIllegalTransition
}

// txQuerier is the minimal subset of *sql.Tx that AdvanceAfterStage
// needs. Defined as an interface so callers can swap a fake in tests
// without spinning up a database.
type txQuerier interface {
	QueryRowContext(ctx context.Context, query string, args ...any) *sql.Row
	ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error)
}

// txBeginner is anything that can open a *sql.Tx. *sql.DB satisfies it.
type txBeginner interface {
	BeginTx(ctx context.Context, opts *sql.TxOptions) (*sql.Tx, error)
}

// AdvanceAfterStage is the SOLE function that mutates videos.state.
// All callers — the orchestrator, the watcher (filesystem trigger),
// Epic 9 (library trigger), Epic 24 (integrity trigger) — funnel
// through here.
//
// Behavior:
//
//  1. Open a transaction and SELECT … FOR UPDATE to serialize
//     concurrent advances on the same row.
//  2. Re-read the current state inside the lock.
//  3. If the current state is a terminal-drop state (SUPERSEDED /
//     CORRUPTED / FAILED) the function logs `late_stage_finish` and
//     commits a no-op. The stage's work is wasted but the state is
//     correct (story edge case 1).
//  4. Look up (from, trigger, outcome). A miss returns
//     *IllegalStateTransition without changing state.
//  5. UPDATE videos.state and updated_at, COMMIT, return the new
//     state. The slot 0004 NOTIFY trigger fires
//     `videos.state_changed` automatically.
//
// The function is dialect-agnostic — the SQL is plain ANSI and works
// against either Postgres or SQLite (via the goose-managed migration
// suite). The pgx-flavored variant in plan-01-06 §7.2 is equivalent;
// we use database/sql here so the package has no third-party deps.
func AdvanceAfterStage(
	ctx context.Context,
	db txBeginner,
	log *slog.Logger,
	videoID string,
	trigger Trigger,
	outcome Outcome,
) (State, error) {
	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		return "", fmt.Errorf("begin tx: %w", err)
	}
	defer func() { _ = tx.Rollback() }()

	state, err := AdvanceInTx(ctx, tx, log, videoID, trigger, outcome)
	if err != nil {
		return state, err
	}
	if err := tx.Commit(); err != nil {
		return "", fmt.Errorf("commit: %w", err)
	}
	return state, nil
}

// AdvanceInTx is the same logic as AdvanceAfterStage but reuses an
// existing transaction. Library merge / integrity sweep callers that
// need to run additional side-effects (e.g. cancel pending probe jobs
// when MISSING → DISCOVERED) take this path so the state flip and the
// side-effect commit atomically.
func AdvanceInTx(
	ctx context.Context,
	tx txQuerier,
	log *slog.Logger,
	videoID string,
	trigger Trigger,
	outcome Outcome,
) (State, error) {
	var current State
	err := tx.QueryRowContext(ctx,
		`SELECT state FROM videos WHERE id = $1 FOR UPDATE`,
		videoID,
	).Scan(&current)
	if err != nil {
		return "", fmt.Errorf("lock video %s: %w", videoID, err)
	}

	// Late-stage finish on a terminal row: log and no-op. Returning
	// the current state (not an error) lets the caller close out
	// cleanly without re-doing the work it already did.
	if current.IsTerminalDrop() {
		if log != nil {
			log.Info("late_stage_finish",
				"video_id", videoID,
				"current", string(current),
				"trigger", string(trigger),
				"outcome", string(outcome),
			)
		}
		return current, nil
	}

	target, ok := Lookup(current, trigger, outcome)
	if !ok {
		return current, &IllegalStateTransition{
			VideoID: videoID,
			From:    current,
			Trigger: trigger,
			Outcome: outcome,
		}
	}

	// The (TRANSCRIBED, *, partial) self-loop is a real edge: we run
	// the UPDATE so updated_at advances and the NOTIFY trigger fires
	// even though state is unchanged. Observers see the partial-gate
	// progress reflected on the row.
	_, err = tx.ExecContext(ctx,
		`UPDATE videos SET state = $1, updated_at = now() WHERE id = $2`,
		string(target), videoID,
	)
	if err != nil {
		return "", fmt.Errorf("update video %s: %w", videoID, err)
	}
	return target, nil
}

File: pipeline/src/maktaba_pipeline/orchestrator/advance.py
"""``advance_after_stage`` — the only sanctioned ``videos.state`` mutator.

Story 1.6 names this function as the single point through which every
state transition flows: stage finishes, filesystem disappearances,
library merges, and integrity-sweep failures all call it. The lint
quarantine in plan-01-06 §11.2 (Python ``ruff`` rule, Go ``forbidigo``
rule) is what mechanically enforces the invariant; this module is the
sanctioned write site.

The function:

1. Opens a transaction and takes a row-level lock on the video (``SELECT
   … FOR UPDATE``).
2. Re-reads the current state inside the lock so concurrent advances
   serialize.
3. If the current state is one of :data:`states.TERMINAL_DROP_STATES`
   (``SUPERSEDED`` / ``CORRUPTED`` / ``FAILED``), logs
   ``late_stage_finish`` and commits a no-op (story edge case 1).
4. Looks up the ``(from, trigger, outcome)`` triple via
   :func:`states.lookup`. A miss raises :class:`IllegalStateTransition`
   without mutating state.
5. UPDATEs ``videos.state`` and ``updated_at``. The slot 0004
   ``videos_state_change_notify_trg`` trigger fires
   ``pg_notify('videos.state_changed', …)`` automatically; on SQLite
   the helper publishes on the in-process bus instead.
"""

from __future__ import annotations

import json
from contextlib import AbstractAsyncContextManager
from typing import Any, Protocol
from uuid import UUID

from ..db.pubsub import get_bus
from ..domain.states import (
    TERMINAL_DROP_STATES,
    IllegalStateTransition,
    Outcome,
    State,
    Trigger,
    lookup,
)

__all__ = ["VIDEOS_STATE_CHANGED", "advance_after_stage"]


# Canonical pubsub channel name for state transitions. Mirrors the
# Postgres NOTIFY channel from slot 0004 so SQLite callers can subscribe
# with the same string.
VIDEOS_STATE_CHANGED = "videos.state_changed"


class _Row(Protocol):
    def __getitem__(self, key: str) -> Any: ...


class _Logger(Protocol):
    def info(self, event: str, **kwargs: Any) -> Any: ...


class DBConn(Protocol):
    """The minimal connection shape :func:`advance_after_stage` needs.

    ``dialect`` is one of ``"postgres"`` or ``"sqlite"``. The Postgres
    NOTIFY trigger fires at the SQL level after the UPDATE commits; the
    SQLite helper publishes manually on the in-process pubsub bus
    because SQLite has no LISTEN/NOTIFY.
    """

    dialect: str

    def transaction(self) -> AbstractAsyncContextManager[Any]: ...

    async def fetchrow(self, sql: str, *args: Any) -> _Row | None: ...

    async def execute(self, sql: str, *args: Any) -> Any: ...


_LOCK_SQL = "SELECT state, library_id FROM videos WHERE id = $1 FOR UPDATE"
_UPDATE_SQL = "UPDATE videos SET state = $1, updated_at = now() WHERE id = $2"


async def advance_after_stage(
    db: DBConn,
    video_id: UUID,
    trigger: Trigger,
    outcome: str | Outcome,
    *,
    log: _Logger,
) -> State:
    """Apply the ``(from, trigger, outcome)`` transition to ``videos``.

    Returns the new state on success, or the unchanged state when the
    row is in a terminal-drop state (a late stage finish — logged as
    ``late_stage_finish``). Raises :class:`IllegalStateTransition` when
    the triple is not in the FSM; the row is left untouched in that
    case.

    The function is dialect-agnostic. On Postgres the NOTIFY trigger
    fires at the SQL layer; on SQLite this function publishes on the
    in-process bus after the transaction commits so subscribers see
    the same payload shape on either backend.
    """
    outcome_value = outcome.value if isinstance(outcome, Outcome) else outcome

    async with db.transaction():
        row = await db.fetchrow(_LOCK_SQL, video_id)
        if row is None:
            raise LookupError(f"video {video_id} not found")
        current = State(row["state"])
        library_id = row["library_id"]

        if current in TERMINAL_DROP_STATES:
            log.info(
                "late_stage_finish",
                video_id=str(video_id),
                current=current.value,
                trigger=trigger.value,
                outcome=outcome_value,
            )
            return current

        target = lookup(current, trigger, outcome_value)
        if target is None:
            raise IllegalStateTransition(video_id, current, trigger, outcome_value)

        await db.execute(_UPDATE_SQL, target.value, video_id)

        # SQLite has no LISTEN/NOTIFY, so the helper publishes manually
        # on the in-process bus. Postgres' AFTER UPDATE trigger handles
        # the equivalent fan-out at the SQL layer.
        if db.dialect == "sqlite":
            get_bus().publish(
                VIDEOS_STATE_CHANGED,
                {
                    "video_id": str(video_id),
                    "library_id": str(library_id) if library_id is not None else None,
                    "old_state": current.value,
                    "new_state": target.value,
                },
            )
        return target


def serialize_notify_payload(
    video_id: UUID,
    library_id: UUID | str | None,
    old_state: State,
    new_state: State,
) -> str:
    """Build the JSON payload the slot 0004 trigger emits.

    Test helper — production code path goes through the SQL trigger.
    Exposed so tests can compare a Python-built payload against what
    Postgres NOTIFY delivers.
    """
    return json.dumps(
        {
            "video_id": str(video_id),
            "library_id": str(library_id) if library_id is not None else None,
            "old_state": old_state.value,
            "new_state": new_state.value,
        },
        separators=(",", ":"),
    )
